jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,570 @@
+	 * @return a method handle bound to the native method
+	 */
+	public static MethodHandle getBoundMethodHandle(Addressable downcallAddr, MethodType functionMethodType, FunctionDescriptor funcDesc) {
+		ProgrammableInvoker nativeInvoker = new ProgrammableInvoker(downcallAddr, functionMethodType, funcDesc);
+		try {
+			MethodHandle boundHandle = lookup.bind(nativeInvoker, "runNativeMethod", methodType(Object.class, long[].class));
+			
+			/* Replace the original handle with the specified types of the C function */
+			boundHandle = nativeInvoker.permuteMH(boundHandle, functionMethodType);
+			return boundHandle;
+		} catch (ReflectiveOperationException e) {
+			throw new InternalError(e);
+		}
+	}
+	
+	/* Collect and convert the passed-in arguments to an Object array for the underlying native call */
+	private MethodHandle permuteMH(MethodHandle targetHandle, MethodType nativeMethodType) throws NullPointerException, WrongMethodTypeException {
+		Class<?>[] argTypeClasses = nativeMethodType.parameterArray();
+		int nativeArgCount = argTypeClasses.length;
+		MethodHandle resultHandle = targetHandle.asCollector(long[].class, nativeArgCount);
+
+		/* Convert the argument values to long via filterArguments() prior to the native call */
+		MethodHandle[] argFilters = new MethodHandle[nativeArgCount];
+		for (int argIndex = 0; argIndex < nativeArgCount; argIndex++) {
+			argFilters[argIndex] = getArgumentFilter(argTypeClasses[argIndex]);
+		}
+		resultHandle = filterArguments(resultHandle, 0, argFilters);
+		
+		/* Convert the return value to the specified type via filterReturnValue() after the native call */
+		MethodHandle retFilter = getReturnValFilter(nativeMethodType.returnType());
+		resultHandle = filterReturnValue(resultHandle, retFilter);
+		return resultHandle;
+	}
+	
+	/* Obtain the filter that converts the passed-in argument to long against its type */
+	private static MethodHandle getArgumentFilter(Class<?> argTypeClass) {
+		/* Set the filter to null in the case of long by default as there is no conversion for long */
+		MethodHandle filterMH = null;
+		
+		if (argTypeClass == boolean.class) {
+			filterMH = booleanToLongArgFilter;
+		} else if (argTypeClass == char.class) {
+			filterMH = charToLongArgFilter;
+		} else if (argTypeClass == byte.class) {
+			filterMH = byteToLongArgFilter;
+		} else if (argTypeClass == short.class) {
+			filterMH = shortToLongArgFilter;
+		} else if (argTypeClass == int.class) {
+			filterMH = intToLongArgFilter;
+		} else if (argTypeClass == float.class) {
+			filterMH = floatToLongArgFilter;
+		} else if (argTypeClass == double.class) {
+			filterMH = doubleToLongArgFilter;
+		} else if (argTypeClass == MemoryAddress.class) {
+			filterMH = memAddrToLongArgFilter;
+		}
+		
+		return filterMH;
+	}
+	
+	/* The return value filter that converts the returned long value from the C function to the specified return type at Java level */
```

Comment:
This line is too long. It's not possible to read it on the web without horizontal scrolling.

Reply:
Fixed as suggested above.